cmd/juju/application/refresher/refresher.go

```diff
@@ -280,3 +351,51 @@ type defaultCharmRepo struct{}
 func (defaultCharmRepo) NewCharmAtPathForceSeries(path, series string, force bool) (charm.Charm, *charm.URL, error) {
 	return charmrepo.NewCharmAtPathForceSeries(path, series, force)
 }
+
+type charmHubRefresher struct {
+	baseRefresher
+}
+
+// Allowed will attempt to check if the charm store is allowed to refresh.
+// Depending on the charm url, will then determine if that's true or not.
+func (r *charmHubRefresher) Allowed(cfg RefresherConfig) (bool, error) {
+	if featureflag.Enabled(feature.CharmHubIntegration) {
+		path, err := charm.EnsureSchema(cfg.CharmRef)
+		if err != nil {
+			return false, errors.Trace(err)
+		}
+
+		curl, err := charm.ParseURL(path)
+		if err != nil {
+			return false, errors.Trace(err)
+		}
+
+		if charm.CharmHub.Matches(curl.Schema) {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+// Refresh a given charm hub charm.
+// Bundles are not supported as there is no physical representation in Juju.
+func (r *charmHubRefresher) Refresh() (*CharmID, error) {
+	newURL, origin, err := r.ResolveCharm()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	curl, _, err := store.AddCharmFromURL(r.charmAdder, newURL, origin, r.force)
```

Comment:
I'm wondering whether ignoring the returned origin is okay or not.  Where do you know that it has the correct revision and channel for the new charm to be install?

Reply:
I have no idea, was going to ask you the exact same question